src/ArtifactUtils.jl

```diff
@@ -70,4 +77,146 @@ function add_artifact!(
     return git_tree_sha1
 end
 
+"""
+    artifact_from_directory(source) -> artifact::SHA1
+
+Create an artifact from the `source` directory and return the `artifact` hash.
+"""
+artifact_from_directory(source) =
+    create_artifact() do artifact_dir
+        cp(source, artifact_dir; force = true, follow_symlinks = true)
+    end
+
+struct GistUploadResult
+    artifact::SHA1
+    filename::String
+    localpath::Union{String,Nothing}
+    url::String
+    sha256::String
+    private::Bool
+end
+
+"""
+    upload_to_gist(
+        artifact::SHA1,
+        [tarball];
+        private::Bool = true,
+        archive_artifact = (),
+        # Following options are aviailable only when `tarball` is not specified:
+        name::AbstractString = "\$artifact.tar.gz",
+        extension::AbstractString = ".tar.gz",
+    ) -> gist
+
+Create an `artifact` archive at path `tarball` (or in a temporary location) and upload it to
+gist. The returned value `gist` can be passed to `add_artifact!`.
+
+# Extended help
+
+## Examples
+```julia
+using ArtifactUtils
+add_artifact!("Artifact.toml", "name", upload_to_gist(artifact_from_directory("source")))
+```
+
+creates an artifact from files in the `"source"` directory, upload it to gist, and then
+add it to `"Artifact.toml"` with the name `"name"`.
+
+## Keyword Arguments
+- `private`: if `true`, upload the archive to a private gist
+- `archive_artifact`: keyword arguments passed to `Pkg.Artifacts.archive_artifact`
+- `name`: name of the archive file, including file extension
+- `extension`: file extension of the tarball. It can be used for specifying the compression
+  method.
+"""
+function upload_to_gist end
+
+function upload_to_gist(
+    artifact::SHA1,
+    tarball::AbstractString;
+    private::Bool = true,
+    archive_artifact = (),
+)
+    mkpath(dirname(tarball))
+    (@__MODULE__).archive_artifact(artifact, tarball; archive_artifact...)
+    sha256 = sha256sum(tarball)
+    url = gist_from_file(tarball; private = private)
+    return GistUploadResult(
+        artifact,
+        basename(tarball),
+        abspath(tarball),
+        url,
+        sha256,
+        private,
+    )
+end
+
+function upload_to_gist(
+    artifact::SHA1;
+    name::Union{AbstractString,Nothing} = nothing,
+    extension::Union{AbstractString,Nothing} = nothing,
+    options...,
+)
```

Comment:
Isn't `upload_to_gist(::SHA1, ::AbstractString)` ambiguous with this definition?

Reply:
This is actually a one-argument function. Maybe you missed the `;` after `SHA1`?

Reply:
Ah yes, you are right!